Current edge 1.18.3 breaks toBuilder on Singular fields with wildcards

The following code compiles with 1.18.2, but not with the current 1.18.3 snapshot (downloaded today) today), using Java 8.

    @Builder(toBuilder = true)
    public class ExampleToBuilder {
        @Singular
        private List<?> objects;
    }

Eclipse says

    Wildcard is not allowed at this location

javac says:

    method emptyList in class java.util.Collections cannot be applied to given types;
    [ERROR]   required: no arguments
    [ERROR]   found: no arguments
    [ERROR]   reason: explicit type argument ? does not conform to declared bound(s) java.lang.Object"

I think 6a42bcc and af55bd0 broke it, but I have not checked that.

--- a/src/core/lombok/eclipse/handlers/HandleBuilder.java
+++ b/src/core/lombok/eclipse/handlers/HandleBuilder.java
@@ -559,7 +559,7 @@ public class HandleBuilder extends EclipseAnnotationHandler<Builder> {
 				MessageSend emptyList = new MessageSend();
 				emptyList.receiver = generateQualifiedNameRef(source, TypeConstants.JAVA, TypeConstants.UTIL, "Collections".toCharArray());
 				emptyList.selector = EMPTY_LIST;
-				emptyList.typeArguments = copyTypes(bfd.singularData.getTypeArgs().toArray(new TypeReference[0]));
+				emptyList.typeArguments = bfd.singularData.getSingularizer().createTypeArgs(bfd.singularData.getTypeArgs().size(), false, type, bfd.singularData.getTypeArgs());
 				ms.arguments = new Expression[] {new ConditionalExpression(ifNull, emptyList, tgt[1])};
 			}
 			ms.receiver = receiver;

--- a/src/core/lombok/eclipse/handlers/HandleBuilder.java
+++ b/src/core/lombok/eclipse/handlers/HandleBuilder.java
@@ -536,7 +536,17 @@ public class HandleBuilder extends EclipseAnnotationHandler<Builder> {
 				boolean obtainIsStatic = bfd.obtainVia.isStatic();
 				for (int i = 0; i < tgt.length; i++) {
 					MessageSend obtainExpr = new MessageSend();
-					obtainExpr.receiver = obtainIsStatic ? new SingleNameReference(type.getName().toCharArray(), 0) : new ThisReference(0, 0);
+					if (obtainIsStatic) {
+						if (typeParams != null && typeParams.length > 0) {
+							obtainExpr.typeArguments = new TypeReference[typeParams.length];
+							for (int j = 0; j<typeParams.length; j++) {
+								obtainExpr.typeArguments[j] = new SingleTypeReference(typeParams[j].name, 0);
+							}
+						}
+						obtainExpr.receiver = new SingleNameReference(type.getName().toCharArray(), 0);
+					} else {
+						obtainExpr.receiver = new ThisReference(0, 0);
+					}
 					obtainExpr.selector = obtainName.toCharArray();
 					if (obtainIsStatic) obtainExpr.arguments = new Expression[] {new ThisReference(0, 0)};
 					tgt[i] = obtainExpr;

--- a/src/core/lombok/javac/handlers/HandleBuilder.java
+++ b/src/core/lombok/javac/handlers/HandleBuilder.java
@@ -506,7 +506,7 @@ public class HandleBuilder extends JavacAnnotationHandler<Builder> {
 				if (bfd.obtainVia.isStatic()) {
 					for (int i = 0; i < tgt.length; i++) {
 						JCExpression c = maker.Select(maker.Ident(type.toName(type.getName())), type.toName(bfd.obtainVia.method()));
-						tgt[i] = maker.Apply(List.<JCExpression>nil(), c, List.<JCExpression>of(maker.Ident(type.toName("this"))));
+						tgt[i] = maker.Apply(typeParameterNames(maker, typeParams), c, List.<JCExpression>of(maker.Ident(type.toName("this"))));
 					}
 				} else {
 					for (int i = 0; i < tgt.length; i++) {

--- a/src/core/lombok/javac/handlers/HandleBuilder.java
+++ b/src/core/lombok/javac/handlers/HandleBuilder.java
@@ -521,7 +521,8 @@ public class HandleBuilder extends JavacAnnotationHandler<Builder> {
 				arg = tgt[0];
 			} else {
 				JCExpression eqNull = maker.Binary(CTC_EQUAL, tgt[0], maker.Literal(CTC_BOT, null));
-				List<JCExpression> tas = cloneTypes(maker, bfd.singularData.getTypeArgs(), ast, type.getContext());
+				// Use the singularizer to create the type args (will remove possible wildcards on the type).
+				List<JCExpression> tas = bfd.singularData.getSingularizer().createTypeArgs(bfd.singularData.getTypeArgs().size(), false, type, bfd.singularData.getTypeArgs(), type.get());
 				JCExpression emptyList = maker.Apply(tas, chainDots(type, "java", "util", "Collections", "emptyList"), List.<JCExpression>nil());
 				arg = maker.Conditional(eqNull, emptyList, tgt[1]);
 			}

--- a/src/core/lombok/javac/handlers/singulars/JavacGuavaSingularizer.java
+++ b/src/core/lombok/javac/handlers/singulars/JavacGuavaSingularizer.java
@@ -169,13 +169,13 @@ abstract class JavacGuavaSingularizer extends JavacSingularizer {
 		List<JCExpression> jceBlank = List.nil();
 		
 		JCExpression varType = chainDotsString(builderType, data.getTargetFqn());
-		int agrumentsCount = getTypeArgumentsCount();
-		varType = addTypeArgs(agrumentsCount, false, builderType, varType, data.getTypeArgs(), source);
+		int argumentsCount = getTypeArgumentsCount();
+		varType = addTypeArgs(argumentsCount, false, builderType, varType, data.getTypeArgs(), source);
 		
 		JCExpression empty; {
 			//ImmutableX.of()
 			JCExpression emptyMethod = chainDots(builderType, "com", "google", "common", "collect", getSimpleTargetTypeName(data), "of");
-			List<JCExpression> invokeTypeArgs = createTypeArgs(agrumentsCount, false, builderType, data.getTypeArgs(), source);
+			List<JCExpression> invokeTypeArgs = createTypeArgs(argumentsCount, false, builderType, data.getTypeArgs(), source);
 			empty = maker.Apply(invokeTypeArgs, emptyMethod, jceBlank);
 		}
 		

--- a/test/transform/resource/after-delombok/BuilderSingularWildcardListsWithToBuilder.java
+++ b/test/transform/resource/after-delombok/BuilderSingularWildcardListsWithToBuilder.java
@@ -1,0 +1,94 @@
+import java.util.List;
+import java.util.Collection;
+class BuilderSingularWildcardListsWithToBuilder {
+	private List<?> objects;
+	private Collection<? extends Number> numbers;
+	@java.lang.SuppressWarnings("all")
+	BuilderSingularWildcardListsWithToBuilder(final List<?> objects, final Collection<? extends Number> numbers) {
+		this.objects = objects;
+		this.numbers = numbers;
+	}
+	@java.lang.SuppressWarnings("all")
+	public static class BuilderSingularWildcardListsWithToBuilderBuilder {
+		@java.lang.SuppressWarnings("all")
+		private java.util.ArrayList<java.lang.Object> objects;
+		@java.lang.SuppressWarnings("all")
+		private java.util.ArrayList<Number> numbers;
+		@java.lang.SuppressWarnings("all")
+		BuilderSingularWildcardListsWithToBuilderBuilder() {
+		}
+		@java.lang.SuppressWarnings("all")
+		public BuilderSingularWildcardListsWithToBuilderBuilder object(final java.lang.Object object) {
+			if (this.objects == null) this.objects = new java.util.ArrayList<java.lang.Object>();
+			this.objects.add(object);
+			return this;
+		}
+		@java.lang.SuppressWarnings("all")
+		public BuilderSingularWildcardListsWithToBuilderBuilder objects(final java.util.Collection<?> objects) {
+			if (this.objects == null) this.objects = new java.util.ArrayList<java.lang.Object>();
+			this.objects.addAll(objects);
+			return this;
+		}
+		@java.lang.SuppressWarnings("all")
+		public BuilderSingularWildcardListsWithToBuilderBuilder clearObjects() {
+			if (this.objects != null) this.objects.clear();
+			return this;
+		}
+		@java.lang.SuppressWarnings("all")
+		public BuilderSingularWildcardListsWithToBuilderBuilder number(final Number number) {
+			if (this.numbers == null) this.numbers = new java.util.ArrayList<Number>();
+			this.numbers.add(number);
+			return this;
+		}
+		@java.lang.SuppressWarnings("all")
+		public BuilderSingularWildcardListsWithToBuilderBuilder numbers(final java.util.Collection<? extends Number> numbers) {
+			if (this.numbers == null) this.numbers = new java.util.ArrayList<Number>();
+			this.numbers.addAll(numbers);
+			return this;
+		}
+		@java.lang.SuppressWarnings("all")
+		public BuilderSingularWildcardListsWithToBuilderBuilder clearNumbers() {
+			if (this.numbers != null) this.numbers.clear();
+			return this;
+		}
+		@java.lang.SuppressWarnings("all")
+		public BuilderSingularWildcardListsWithToBuilder build() {
+			java.util.List<java.lang.Object> objects;
+			switch (this.objects == null ? 0 : this.objects.size()) {
+			case 0: 
+				objects = java.util.Collections.emptyList();
+				break;
+			case 1: 
+				objects = java.util.Collections.singletonList(this.objects.get(0));
+				break;
+			default: 
+				objects = java.util.Collections.unmodifiableList(new java.util.ArrayList<java.lang.Object>(this.objects));
+			}
+			java.util.Collection<Number> numbers;
+			switch (this.numbers == null ? 0 : this.numbers.size()) {
+			case 0: 
+				numbers = java.util.Collections.emptyList();
+				break;
+			case 1: 
+				numbers = java.util.Collections.singletonList(this.numbers.get(0));
+				break;
+			default: 
+				numbers = java.util.Collections.unmodifiableList(new java.util.ArrayList<Number>(this.numbers));
+			}
+			return new BuilderSingularWildcardListsWithToBuilder(objects, numbers);
+		}
+		@java.lang.Override
+		@java.lang.SuppressWarnings("all")
+		public java.lang.String toString() {
+			return "BuilderSingularWildcardListsWithToBuilder.BuilderSingularWildcardListsWithToBuilderBuilder(objects=" + this.objects + ", numbers=" + this.numbers + ")";
+		}
+	}
+	@java.lang.SuppressWarnings("all")
+	public static BuilderSingularWildcardListsWithToBuilderBuilder builder() {
+		return new BuilderSingularWildcardListsWithToBuilderBuilder();
+	}
+	@java.lang.SuppressWarnings("all")
+	public BuilderSingularWildcardListsWithToBuilderBuilder toBuilder() {
+		return new BuilderSingularWildcardListsWithToBuilderBuilder().objects(this.objects == null ? java.util.Collections.<java.lang.Object>emptyList() : this.objects).numbers(this.numbers == null ? java.util.Collections.<Number>emptyList() : this.numbers);
+	}	
+}

--- a/test/transform/resource/after-delombok/BuilderWithToBuilder.java
+++ b/test/transform/resource/after-delombok/BuilderWithToBuilder.java
@@ -86,7 +86,7 @@ class BuilderWithToBuilder<T> {
 	}
 	@java.lang.SuppressWarnings("all")
 	public BuilderWithToBuilderBuilder<T> toBuilder() {
-		return new BuilderWithToBuilderBuilder<T>().one(this.mOne).two(this.mTwo).foo(BuilderWithToBuilder.rrr(this)).bars(this.bars == null ? java.util.Collections.<T>emptyList() : this.bars);
+		return new BuilderWithToBuilderBuilder<T>().one(this.mOne).two(this.mTwo).foo(BuilderWithToBuilder.<T>rrr(this)).bars(this.bars == null ? java.util.Collections.<T>emptyList() : this.bars);
 	}
 }
 class ConstructorWithToBuilder<T> {

--- a/test/transform/resource/after-ecj/BuilderSingularWildcardListsWithToBuilder.java
+++ b/test/transform/resource/after-ecj/BuilderSingularWildcardListsWithToBuilder.java
@@ -1,0 +1,87 @@
+import java.util.List;
+import java.util.Collection;
+import lombok.Singular;
+@lombok.Builder(toBuilder = true) class BuilderSingularWildcardListsWithToBuilder {
+  public static @java.lang.SuppressWarnings("all") class BuilderSingularWildcardListsWithToBuilderBuilder {
+    private @java.lang.SuppressWarnings("all") java.util.ArrayList<java.lang.Object> objects;
+    private @java.lang.SuppressWarnings("all") java.util.ArrayList<Number> numbers;
+    @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder() {
+      super();
+    }
+    public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder object(final java.lang.Object object) {
+      if ((this.objects == null))
+          this.objects = new java.util.ArrayList<java.lang.Object>();
+      this.objects.add(object);
+      return this;
+    }
+    public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder objects(final java.util.Collection<?> objects) {
+      if ((this.objects == null))
+          this.objects = new java.util.ArrayList<java.lang.Object>();
+      this.objects.addAll(objects);
+      return this;
+    }
+    public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder clearObjects() {
+      if ((this.objects != null))
+          this.objects.clear();
+      return this;
+    }
+    public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder number(final Number number) {
+      if ((this.numbers == null))
+          this.numbers = new java.util.ArrayList<Number>();
+      this.numbers.add(number);
+      return this;
+    }
+    public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder numbers(final java.util.Collection<? extends Number> numbers) {
+      if ((this.numbers == null))
+          this.numbers = new java.util.ArrayList<Number>();
+      this.numbers.addAll(numbers);
+      return this;
+    }
+    public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder clearNumbers() {
+      if ((this.numbers != null))
+          this.numbers.clear();
+      return this;
+    }
+    public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilder build() {
+      java.util.List<java.lang.Object> objects;
+      switch (((this.objects == null) ? 0 : this.objects.size())) {
+      case 0 :
+          objects = java.util.Collections.emptyList();
+          break;
+      case 1 :
+          objects = java.util.Collections.singletonList(this.objects.get(0));
+          break;
+      default :
+          objects = java.util.Collections.unmodifiableList(new java.util.ArrayList<java.lang.Object>(this.objects));
+      }
+      java.util.Collection<Number> numbers;
+      switch (((this.numbers == null) ? 0 : this.numbers.size())) {
+      case 0 :
+          numbers = java.util.Collections.emptyList();
+          break;
+      case 1 :
+          numbers = java.util.Collections.singletonList(this.numbers.get(0));
+          break;
+      default :
+          numbers = java.util.Collections.unmodifiableList(new java.util.ArrayList<Number>(this.numbers));
+      }
+      return new BuilderSingularWildcardListsWithToBuilder(objects, numbers);
+    }
+    public @java.lang.Override @java.lang.SuppressWarnings("all") java.lang.String toString() {
+      return (((("BuilderSingularWildcardListsWithToBuilder.BuilderSingularWildcardListsWithToBuilderBuilder(objects=" + this.objects) + ", numbers=") + this.numbers) + ")");
+    }
+  }
+  private @Singular List<?> objects;
+  private @Singular Collection<? extends Number> numbers;
+  @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilder(final List<?> objects, final Collection<? extends Number> numbers) {
+    super();
+    this.objects = objects;
+    this.numbers = numbers;
+  }
+  public static @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder builder() {
+    return new BuilderSingularWildcardListsWithToBuilderBuilder();
+  }
+  public @java.lang.SuppressWarnings("all") BuilderSingularWildcardListsWithToBuilderBuilder toBuilder() {
+    return new BuilderSingularWildcardListsWithToBuilderBuilder().objects(((this.objects == null) ? java.util.Collections.<java.lang.Object>emptyList() : this.objects)).numbers(((this.numbers == null) ? java.util.Collections.<Number>emptyList() : this.numbers));
+  }
+}

--- a/test/transform/resource/after-ecj/BuilderWithToBuilder.java
+++ b/test/transform/resource/after-ecj/BuilderWithToBuilder.java
@@ -74,7 +74,7 @@ import lombok.Builder;
     return new BuilderWithToBuilderBuilder<T>();
   }
   public @java.lang.SuppressWarnings("all") BuilderWithToBuilderBuilder<T> toBuilder() {
-    return new BuilderWithToBuilderBuilder<T>().one(this.mOne).two(this.mTwo).foo(BuilderWithToBuilder.rrr(this)).bars(((this.bars == null) ? java.util.Collections.<T>emptyList() : this.bars));
+    return new BuilderWithToBuilderBuilder<T>().one(this.mOne).two(this.mTwo).foo(BuilderWithToBuilder.<T>rrr(this)).bars(((this.bars == null) ? java.util.Collections.<T>emptyList() : this.bars));
   }
 }
 @lombok.experimental.Accessors(prefix = "m") class ConstructorWithToBuilder<T> {

--- a/test/transform/resource/before/BuilderSingularWildcardListsWithToBuilder.java
+++ b/test/transform/resource/before/BuilderSingularWildcardListsWithToBuilder.java
@@ -1,0 +1,10 @@
+import java.util.List;
+import java.util.Collection;
+
+import lombok.Singular;
+
+@lombok.Builder(toBuilder = true)
+class BuilderSingularWildcardListsWithToBuilder {
+	@Singular private List<?> objects;
+	@Singular private Collection<? extends Number> numbers;
+}

